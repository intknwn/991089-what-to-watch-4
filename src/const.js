export const STATIC_RESOURSES_URL = `https://4.react.pages.academy`;

export const MAX_GENRES_COUNTER = 9;
export const MAX_MORE_LIKE_THIS_MOVIES_COUNTER = 4;
export const MOVIES_PER_PAGE = 8;

export const MOVIES_LIST_PLAYER_CONFIG = {
  autoPlay: true,
  muted: true,
  loop: true,
};

export const MOVIE_PAGE_PLAYER_CONFIG = {
  autoPlay: false,
  muted: false,
  loop: false,
};

export const Tab = {
  OVERVIEW: `Overview`,
  DETAILS: `Details`,
  REVIEWS: `Reviews`,
};

export const Genre = {
  ALL_GENRES: `All genres`,
};

export const AuthStatus = {
  AUTH: `AUTH`,
  NOT_AUTH: `NOT_AUTH`,
};

export const HttpCode = {
  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  INTERNAL_ERROR: 500,
  SERVICE_UNAVAILABLE: 503,
};

export const ReviewForm = {
  MIN_REVIEW_LENGTH: 50,
  MAX_REVIEW_LENGTH: 400,
  MAX_RATE: 5,
};

export const AppRoute = {
  MAIN: `/`,
  SIGN_IN: `/sign-in`,
  MY_LIST: `/mylist`,
  REVIEW: `/review`,
  PLAYER: `/player`,
  MOVIE: `/movies`,
};

export const ClassName = {
  MOVIE_PAGE_HEADER: `page-header movie-card__head`,
  USER_PAGE_HEADER: `page-header user-page__head`,
  LOGO_LINK: `logo__link`,
  LOGO_LINK_LIGHT: `logo__link logo__link--light`,
};
