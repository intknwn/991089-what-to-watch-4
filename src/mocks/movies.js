export default [
  {
    name: `Fantastic Beasts: The Crimes of Grindelwald`,
    previewImg: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Fantasy`,
    year: 2018,
    rating: 6.6,
    score: 240,
    description: `In an effort to thwart Grindelwald's plans of raising pure-blood wizards to rule over all non-magical beings, Albus Dumbledore enlists his former student Newt Scamander, who agrees to help, unaware of the dangers that lie ahead. Lines are drawn as love and loyalty are tested, even among the truest friends and family, in an increasingly divided wizarding world.`,
    director: `David Yates`,
    cast: [`Eddie Redmayne`, `Katherine Waterston`, `Dan Fogler`],
    runtime: 134,
  },
  {
    name: `Bohemian Rhapsody`,
    previewImg: `img/bohemian-rhapsody.jpg`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `Biography`,
    year: 2018,
    rating: 8.0,
    score: 450,
    description: `The story of the legendary British rock band Queen and lead singer Freddie Mercury, leading up to their famous performance at Live Aid (1985).`,
    director: `Bryan Singer`,
    cast: [`Rami Malek`, `Lucy Boynton`, `Gwilym Lee`],
    runtime: 134,
  },
  {
    name: `We need to talk about Kevin`,
    previewImg: `img/we-need-to-talk-about-kevin.jpg`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Thriller`,
    year: 2011,
    rating: 7.5,
    score: 138,
    description: `Kevin's mother struggles to love her strange child, despite the increasingly dangerous things he says and does as he grows up. But Kevin is just getting started, and his final act will be beyond anything anyone imagined.`,
    director: `Lynne Ramsay`,
    cast: [`Tilda Swinton`, `John C. Reilly`, `Ezra Miller`],
    runtime: 112,
  },
  {
    name: `What We Do in the Shadows`,
    previewImg: `img/what-we-do-in-the-shadows.jpg`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `Comedy`,
    year: 2019,
    rating: 8.5,
    score: 36894,
    description: `A look into the daily (or rather, nightly) lives of three vampires, who've lived together for over 100 years, on Staten Island.`,
    director: `Jemaine Clement`,
    cast: [`Kayvan Novak`, `Matt Berry`, `Natasia Demetriou`],
    runtime: 86,
  },
  {
    name: `Revenant`,
    previewImg: `img/revenant.jpg`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
    genre: `Adventure`,
    year: 2015,
    rating: 8.0,
    score: 710693,
    description: `A frontiersman on a fur trading expedition in the 1820s fights for survival after being mauled by a bear and left for dead by members of his own hunting team.`,
    director: `Alejandro G. Iñárritu`,
    cast: [`Leonardo DiCaprio`, `Tom Hardy`, `Will Poulter`],
    runtime: 156,
  },
  {
    name: `Johnny English`,
    previewImg: `img/johnny-english.jpg`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Comedy`,
    year: 2003,
    rating: 6.2,
    score: 149199,
    description: `After a sudden attack on MI5, Johnny English, Britain's most confident, yet unintelligent spy, becomes Britain's only spy.`,
    director: `Peter Howitt`,
    cast: [`Rowan Atkinson`, `John Malkovich`, `Natalie Imbruglia`],
    runtime: 89,
  },
  {
    name: `Falling Down`,
    previewImg: `https://via.placeholder.com/280x175`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Crime`,
    year: 1993,
    rating: 7.6,
    score: 173264,
    description: `An ordinary man frustrated with the various flaws he sees in society begins to psychotically and violently lash out against them.`,
    director: `Joel Schumacher`,
    cast: [`Michael Douglas`, `Robert Duvall`, `Barbara Hershey`],
    runtime: 113,
  },
  {
    name: `Batman & Robin`,
    previewImg: `https://via.placeholder.com/280x175`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Action`,
    year: 1997,
    rating: 3.8,
    score: 235550,
    description: `Batman and Robin try to keep their relationship together even as they must stop Mr. Freeze and Poison Ivy from freezing Gotham City.`,
    director: `Joel Schumacher`,
    cast: [`Arnold Schwarzenegger`, `George Clooney`, `Chris O'Donnell`],
    runtime: 125,
  },
  {
    name: `Terminator 2: Judgment Day`,
    previewImg: `https://via.placeholder.com/280x175`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Action`,
    year: 1991,
    rating: 8.5,
    score: 1002326,
    description: `A cyborg, identical to the one who failed to kill Sarah Connor, must now protect her ten year old son, John Connor, from a more advanced and powerful cyborg.`,
    director: `James Cameron`,
    cast: [`Arnold Schwarzenegger`, `Linda Hamilton`, `Edward Furlong`],
    runtime: 125,
  },
  {
    name: `Blade Runner 2049`,
    previewImg: `https://via.placeholder.com/280x175`,
    backgroundImg: `https://via.placeholder.com/1300x552`,
    posterImg: `https://via.placeholder.com/273x410`,
    previewVid: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`,
    genre: `Action`,
    year: 2017,
    rating: 8.0,
    score: 469353,
    description: `Young Blade Runner K's discovery of a long-buried secret leads him to track down former Blade Runner Rick Deckard, who's been missing for thirty years.`,
    director: `Denis Villeneuve`,
    cast: [`Harrison Ford`, `Ryan Gosling`, `Ana de Armas`],
    runtime: 164,
  },
];
