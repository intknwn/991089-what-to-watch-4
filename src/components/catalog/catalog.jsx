import React from 'react';
import {arrayOf, func, string} from 'prop-types';
import {GenresList, MoviesList} from '../components.jsx';
import {withShowMore} from '../../hocs/hocs.js';
import {movieType} from '../../types/types.js';
import {Genre, MOVIES_PER_PAGE} from '../../const.js';

const MoviesListWrapped = withShowMore(MoviesList);

class Catalog extends React.PureComponent {
  constructor(props) {
    super(props);

    this.isDefault = true;
  }

  componentDidUpdate(prevProps) {
    if (this.props.activeItem !== prevProps.activeItem) {

      this.isDefault = !this.isDefault;
    }
  }

  render() {
    const {
      activeItem: genre,
      activeItemChangeHandler: onGenreClick,
      genres,
      movies,
    } = this.props;

    const selectedByGenreMovies = genre && genre !== Genre.ALL_GENRES ?
      movies.filter((movie) => movie.genre === genre) :
      movies;

    return (
      <section className="catalog">
        <h2 className="catalog__title visually-hidden">Catalog</h2>
        <GenresList
          genres={genres}
          selectedGenre={genre}
          onGenreClick={onGenreClick}
        />
        <MoviesListWrapped
          movies={selectedByGenreMovies}
          isDefault={this.isDefault}
          cb={(items) => items + MOVIES_PER_PAGE}
        />
      </section>
    );
  }
}

Catalog.propTypes = {
  activeItem: string.isRequired,
  activeItemChangeHandler: func.isRequired,
  genres: arrayOf(string).isRequired,
  movies: arrayOf(movieType).isRequired,
};

export default Catalog;
