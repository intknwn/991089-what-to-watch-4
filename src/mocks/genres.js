import {Genre} from '../const.js';

export default [
  Genre.ALL_GENRES,
  `Fantasy`,
  `Biography`,
  `Thriller`,
  `Comedy`,
  `Adventure`,
  `Crime`,
  `Action`,
];
