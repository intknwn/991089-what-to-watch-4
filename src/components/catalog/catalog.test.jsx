import React from 'react';
import renderer from 'react-test-renderer';
import {Catalog} from './catalog.jsx';
import {Genre} from '../../const.js';
import genres from '../../mocks/genres.js';
import movies from '../../mocks/movies.js';

it(`Catalog component should render movies catalog`, () => {
  const tree = renderer
      .create(
          <Catalog
            genres={genres}
            selectedGenre={Genre.ALL_GENRES}
            onGenreClick={() => {}}
            selectedByGenreMovies={movies}
            isShowMore={true}
            onMovieCardClickHandler={() => {}}
            onShowMoreClick={() => {}}
          />
      )
      .toJSON();

  expect(tree).toMatchSnapshot();
});
