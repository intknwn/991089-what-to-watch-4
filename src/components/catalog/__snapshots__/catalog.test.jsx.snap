// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Catalog component should render movies catalog 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Fantasy
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Biography
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Thriller
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Comedy
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Adventure
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Crime
      </a>
    </li>
    <li
      className="catalog__genres-item"
    >
      <a
        className="catalog__genres-link"
        href="#"
        onClick={[Function]}
      >
        Action
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="Fantastic Beasts: The Crimes of Grindelwald"
          height={175}
          src="img/fantastic-beasts-the-crimes-of-grindelwald.jpg"
          width={280}
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/movies/1"
          onClick={[Function]}
        >
          Fantastic Beasts: The Crimes of Grindelwald
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="Bohemian Rhapsody"
          height={175}
          src="img/bohemian-rhapsody.jpg"
          width={280}
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/movies/2"
          onClick={[Function]}
        >
          Bohemian Rhapsody
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="We need to talk about Kevin"
          height={175}
          src="img/we-need-to-talk-about-kevin.jpg"
          width={280}
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/movies/3"
          onClick={[Function]}
        >
          We need to talk about Kevin
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onClick={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
      >
        <img
          alt="What We Do in the Shadows"
          height={175}
          src="img/what-we-do-in-the-shadows.jpg"
          width={280}
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/movies/4"
          onClick={[Function]}
        >
          What We Do in the Shadows
        </a>
      </h3>
    </article>
  </div>
</section>
`;
